src/crocodilehunter.py

```diff
@@ -101,6 +100,27 @@ def start(self):
         if return_code:
             self.logger.critical("Bootstrapping failed")
             self.cleanup(True)
+        
+        # Test GPSD
+        if self.disable_gps:
+            self.logger.info("GPS disabled. Skipping test.")
+        else:
+            gpsd.connect()
+            packet = gpsd.get_current()
+            if packet.mode > 1:
+                self.logger.info("GPS fix detected")
+            else:
+                self.logger.info("No GPS fix detected. Waiting for fix.")
+                packet = gpsd.get_current()
+                tries = 1
+                while packet.mode < 2:
+                    if tries < 10:
+                        self.logger.debug("Try %s waiting for fix.")
+                        packet = gpsd.get_current()
+                        tries += 1
+                    else:
+                        self.logger.critical("No GPS fix detected. Exiting.")
+                        exit(1)
```

Comment:
This all looks good and when it works it seems to work well. But when gpsd is not running or can't connect I get the following ugly error
```
----------------------------------------
Exception happened during processing of request from 
Traceback (most recent call last):
  File "/usr/lib/python3.8/socketserver.py", line 316, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python3.8/socketserver.py", line 347, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python3.8/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python3.8/socketserver.py", line 720, in __init__
    self.handle()
  File "/home/cooperq/Code/crocodilehunter/src/watchdog.py", line 542, in handle
    wd_inst.process_tower(data)
  File "/home/cooperq/Code/crocodilehunter/src/watchdog.py", line 243, in process_tower
    packet = self.get_gps()
  File "/home/cooperq/Code/crocodilehunter/src/watchdog.py", line 226, in get_gps
    gpsd.connect(host='127.0.0.1', port=2948)
  File "/home/cooperq/.local/lib/python3.8/site-packages/gpsd/__init__.py", line 254, in connect
    gpsd_socket.connect((host, port))
ConnectionRefusedError: [Errno 111] Connection refused
----------------------------------------
```
You may also notice that from the above error I have to specify the host and port for my GPSD because it's running on a non standard port (there is a long explanation for why this is.) 
This combined with the other problems I mentioned above in the bootstrap script lead me to the conclusion that we shouldn't be relying on gpsd but perhaps instead we should just be parsing the GPS info directly from the device itself. There is a guide here on how to do this in python2 : https://ozzmaker.com/using-python-with-a-gps-receiver-on-a-raspberry-pi/
I was able to port this to python 3 for an experiment pretty easily and I think we should replace the gpsd import with this.

Reply:
I think we have a few things here:
**Non-default GPSD host/port:** Maybe we add an optional configuration value that allows you to set the GPSD host and port in the event someone wants or needs it on a different host or port. If it's not specified we could use some sensible defaults like 127.0.0.1 and the default GPSD port.
**Ugly stack trace:** I should probably add some exception checking for common problems as well as documentation hints for specific errors in the error message itself.
**Library that grabs the entire GPS device:** I don't like doing that because I've created applications which leverage libraries that do a similar thing and it prevents other applications from interfacing with the GPS. I don't think we should assume our application is the only thing that should have access to a given GPS device in a given environment. Using the GPSD suite of applications and libraries allows us to play nicely with other applications and also means it's possible to leverage networked GPSes as well as GPSD's suite of test programs and utilities which are handy for debugging, development, and demonstrations.

Let me know what you think about all these.

Reply:
Yea okay I see what you are saying about playing nice with other applications and not grabbing the GPS device. If we stick with GPSd then I think the right thing to do is make the host/port/device configurable, make sure gpsd is running when we start (and start it if not), and document how to set up GPSD,